roles/kustomize_deploy/tasks/execute_step.yml

```diff
@@ -0,0 +1,189 @@
+             stage['path'], item.src_file) | path_join
+          }}
+        get_attributes: false
+        get_checksum: false
+        get_mime: false
+      loop: "{{ stage['values'] }}"
+
+    - name: Assert source files exist
+      ansible.builtin.assert:
+        that:
+          - (item.stat is defined and item.stat.exists) or item.skipped
+        quiet: true
+      loop: "{{ _src.results }}"
+      loop_control:
+        label: "{{ item.stat.path | default('none') | basename }}"
+
+    - name: "Executing pre_deploy step for {{ stage.path }}"
+      when:
+        - not cifmw_kustomize_deploy_generate_crs_only | bool
+        - stage.pre_stage_run is defined
+      vars:
+        hooks: "{{ stage.pre_stage_run }}"
+        step: "pre_stage_run_{{ stage_id }}"
+      ansible.builtin.include_role:
+        name: run_hook
+
+    - name: "Generate values.yaml for {{ stage.path }}"
+      when:
+        - _val.src_file is not none
+        - _val.src_file != 'none'
+      vars:
+        cifmw_ci_gen_kustomize_values_name: >-
+          {{ _val['name'] }}
+        cifmw_ci_gen_kustomize_values_dt_src_file: >-
+          {{
+            (cifmw_kustomize_deploy_architecture_repo_dest_dir,
+             stage['path'], _val.src_file
+            ) |
+            path_join
+           }}
+        cifmw_ci_gen_kustomize_values_userdata: >-
+          {{
+            (cifmw_architecture_user_kustomize is defined and
+             cifmw_architecture_user_kustomize['stage_' ~ stage_id] is defined
+            ) | ternary(cifmw_architecture_user_kustomize['stage_' ~ stage_id],
+                        {})
+          }}
+      ansible.builtin.include_role:
+        name: ci_gen_kustomize_values
+      loop: "{{ stage['values'] }}"
+      loop_control:
+        loop_var: _val
+        index_var: _val_id
+        label: "{{ _val['name'] }}"
+
+    - name: "Copy generated values for {{ stage.path }}"
+      when:
+        - _val.src_file is not none
+        - _val.src_file != 'none'
+      ansible.builtin.copy:
+        backup: true
+        remote_src: true
+        src: >-
+          {{
+            (cifmw_kustomize_deploy_basedir,
+             'artifacts', 'ci_gen_kustomize_values',
+             _val['name'], 'values.yaml') | path_join
+          }}
+        dest: >-
+          {{
+            (cifmw_kustomize_deploy_architecture_repo_dest_dir,
+            stage['path'], _val.src_file
+            ) |
+            path_join
+           }}
+      loop: "{{ stage['values'] }}"
+      loop_control:
+        loop_var: _val
+        label: "{{ _val['name'] }}"
+
+    - name: Stop before building kustomization if requested
+      when:
+        - cifmw_deploy_va_stopper is defined
+        - cifmw_deploy_va_stopper == _stage_stopper
+      vars:
+        _stage_stopper: "pre_kustomize_stage_{{ stage_id }}"
+      ansible.builtin.fail:
+        msg: "Failing on demand {{ cifmw_deploy_va_stopper }}"
```

Comment:
I know we already talked about this, and that this may be the best possible approach. I have been checking how the `end_play` and `end_host` meta works and based on the code I think it works at play level, meaning, in this case, that calling it inside a loop ruins the entire purpose. Have you tried to do the following?

1. Remove the fail call
2. Call the `ansible.builtin.meta` with the `end_play` argument when needed and set an internal fact that signal that we want to stop (non-cacheable).
3. Add it as a condition in every call. We already have a when clause in every task here so I don't think its a big deal.
4. In the upper caller loop, add that flag as a condition. As the loop is in a playbook I'm pretty sure the iterator that handles it will properly stop.

Reply:
so I'm 100% sure `end_host` doesn't work. I'll do another test with `end_play` - but I have little to no confidence it will really stop the whole playbook.

Also, there is a need to have an actual "break point" in the loop itself - being before or after some specific parts. Removing that capability will just call for "hey we need this to debug that specific part" - so I really don't think removing it is a good idea.
Moving the end_play to the upper level (and having some more `when` in the tasks) would just make the whole code more complicated to maintain imho.

Reply:
So, I did some more tests here:
- define a fact before the loop
- start of the loop is checking if the defined fact exists, and if its value (true|false)
- during the loop, move tasks in blocks with `when` conditions based on an iterative approach (easier to manage)
- last task in the loop is to update the fact defined before the loop, in case we want to stop
- run, run, run
Result:
- the 06-deploy-va.yml "main" playbook is indeed stopped (both with `end_play` and `end_host`)
- but then it goes straight to the next playbook we import (hooks, then the rest of the list as set in the deploy-edpm.yml)
- eventually fails because of missing data, since we didn't finish to deploy OpenStack itself

sooo.... apparently, failing is the "best" way to actually stop the whole process....

Reply:
Nothing to do there then: we can't do it better without hacking really deep in some custom plugin.